### ISSUE TITLE
[icetray] frame key size (Trac #2042)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#2034]
For some reason, the I3Frame size for each key gets wiped out when it gets loaded.  While I understand that it's just returning the backing blob size, and that went away when we loaded it, this was extremely convenient for knowing how large an object was.

```text
In [7]: fr2.size('I3Geometry')
Out[7]: 618296

In [8]: fr2['I3Geometry']
Out[8]: <icecube.dataclasses.I3Geometry at 0x7f4fa83118c0>

In [9]: fr2.size('I3Geometry')
Out[9]: 0
```

Can we implement a cached value of this in I3Frame?  Previously we had a hack in the shovels that wrapped the frame and grabbed all the values before they disappeared.  But this disappeared with the work in #1423.

Migrated from https://code.icecube.wisc.edu/ticket/2042
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:13:57", 
    "description": "For some reason, the I3Frame size for each key gets wiped out when it gets loaded.  While I understand that it's just returning the backing blob size, and that went away when we loaded it, this was extremely convenient for knowing how large an object was.\n\n{{{\nIn [7]: fr2.size('I3Geometry')\nOut[7]: 618296\n\nIn [8]: fr2['I3Geometry']\nOut[8]: <icecube.dataclasses.I3Geometry at 0x7f4fa83118c0>\n\nIn [9]: fr2.size('I3Geometry')\nOut[9]: 0\n}}}\n\nCan we implement a cached value of this in I3Frame?  Previously we had a hack in the shovels that wrapped the frame and grabbed all the values before they disappeared.  But this disappeared with the work in #1431.", 
    "reporter": "david.schultz", 
    "cc": "olivas", 
    "resolution": "fixed", 
    "_ts": "1550067237750774", 
    "component": "combo core", 
    "summary": "[icetray] frame key size", 
    "priority": "critical", 
    "keywords": "", 
    "time": "2017-06-21T22:55:41", 
    "milestone": "", 
    "owner": "david.schultz", 
    "type": "enhancement"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20170621T23:16:54: cweaver commented:

> It looks like the reason this information is lost is that we are relying on the the `blob_t`'s `buf` member to store it for us, but since that is a `vector` it has the invariant that `size <= capacity`, and we wipe out the capacity in order to avoid paying ~double the memory cost for each frame object. So, to continue providing a stale version of this information (which can only be used as a guide if the object may have been modified) we would need to add another size field to `blob_t` to hold it. This probably isn't a concern memory use-wsie, although I feel a little unhappy having to spend an extra 8 bytes per object for something which has no direct use. I guess if we really cared we could use a union, however, since this field would never be used at the same time as `buf`.

**Comment by icecube-trac** [comment]
Trac update at 20170622T21:43:07: 
* david.schultz changed resolution from "" to "fixed"
* david.schultz changed status from "new" to "closed"
* david.schultz commented:

> In changeset 2921:
> 
> ```text
> #!CommitTicketReference repository="IceTray" revision="2921"
> keep serialized sizes around in a cached value. closes #2034
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190111T20:59:34: nwhitehorn commented:

> In changeset 168822:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="168822"
> keep serialized sizes around in a cached value. closes #2034
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:13:57: 
* olivas changed milestone from "June 2017" to ""
* olivas commented:

> Milestone June 2017 deleted
